design/one-pager-package-image-config.md

```diff
@@ -0,0 +1,377 @@
+      - `name`: The name of the authority.
+      - `keyless`: Keyless verification settings.
+        - `url`: The URL of the keyless authority.
+        - `identities`: List of identities to use for verification.
+          - `issuer`: The issuer of the identity.
+          - `subjectRegExp`: The regular expression to match the subject of the
+            identity.
+      - `attestations`: List of attestations to use for verification.
+        - `name`: The name of the attestation.
+        - `predicateType`: The type of the predicate to use for verification.
+
+The `ImageConfig` object will be a global object in the cluster, configuring
+the package manager behavior without being referenced by any other object. There
+could be multiple `ImageConfig` objects in the cluster, each defining settings
+for different sets of images.
+
+When a package image needs to be pulled, either as a primary package or as a
+dependency, the package manager will search for `ImageConfig` objects in the
+cluster that match the image. The package manager selects the best match based
+on the longest prefix match with the image. If there are multiple matches with
+the same prefix length (i.e., identical prefixes across `ImageConfig`s), the
+package manager will select one arbitrarily. Similarly, after the image is
+pulled, the package manager will query the `ImageConfig` objects to find the
+best matching verification settings for the image. The selected `ImageConfig`
+may be different for authentication and verification settings if there are
+separate objects defined for these settings.
+
+Careful readers might have noticed that the `spec.verification.cosign` field
+closely follows the schema used in the _Policy Controller's_ `ClusterImagePolicy`
+API. This is a deliberate design choice to ensure the API is flexible enough to
+handle various image verification setups while also providing a consistent user
+experience for those familiar with the Policy Controller. Since both _Policy
+Controller_ and _Cosign_ are developed by the same organization, we believe
+there's no better source of expertise for verifying Cosign-signed images. We
+plan to leverage this expertise, along with existing libraries from the Policy
+Controller project, to implement reliable image verification in the package
+manager.
+
+### User Experience
+
+We anticipate that the following would be important for users of this API:
+
+- To understand which `ImageConfig` object would be selected for a given package.
+- To understand whether image verification is skipped, succeeded, or failed for
+  a given package image.
+
+Considering this API primarily impacts the package revision APIs, namely
+`ProviderRevision`, `ConfigurationRevision`, and `FunctionRevision`, we believe
+that it is a good idea to communicate these details on those objects. The selected
+`ImageConfig` object will be communicated as an event on the revision object.
```

Comment:
As long as we don't spam these every reconcile loop fine, I could see a condition for the `SelectedImageConfig` too.

Reply:
It is possible to select one ImageConfig for auth (e.g. pull secret) but another one for signature verification. So, I am planning to emit dedicated events as in the given example in the design. In the future, there could be others, e.g. registry tls, and it doesn't feel scalable to have dedicated conditions for each and reusing the same condition for all would be a little painful.

Reply:
this means we are ruling out erroring out on multiple matches, right?

Reply:
See https://github.com/crossplane/crossplane/pull/5993#discussion_r1795461707